src/backend_api.cpp

```diff
@@ -0,0 +1,268 @@
+#include <miopen/backend_api.hpp>
+#include <miopen/miopen.h>
+#include <miopen/errors.hpp>
+#include <new>
+
+extern "C" miopenStatus_t
+miopenBackendCreateDescriptor(miopenBackendDescriptorType_t descriptorType,
+                              miopenBackendDescriptor_t* descriptor)
+{
+    switch(descriptorType)
+    {
+    case MIOPEN_BACKEND_TENSOR_DESCRIPTOR:
+        // TODO: add logging
+        // TODO: get rid of unnecessary throws
+        return miopen::try_(
+            [&] {
+                miopen::backend_api::BackendTensorDescriptor* newDescriptor =
+                    new(std::nothrow) miopen::backend_api::BackendTensorDescriptor();
+
+                if(!newDescriptor)
+                {
+                    throw miopen::Exception(miopenStatusAllocFailed);
+                }
+
+                try
+                {
+                    if(newDescriptor->Status() != miopenStatusSuccess)
+                    {
+                        miopenStatus_t status = newDescriptor->Status();
+                        throw miopen::Exception(status);
+                    }
+
+                    miopen::deref(descriptor) = newDescriptor;
+                }
+                catch(...)
+                {
+                    delete newDescriptor;
+                    throw;
+                }
+            },
+            false);
+
+    default: return miopenStatus_t::miopenStatusUnsupportedOp;
+    }
+}
+
+extern "C" miopenStatus_t miopenBackendSetAttribute(miopenBackendDescriptor_t descriptor,
+                                                    miopenBackendAttributeName_t attributeName,
+                                                    miopenBackendAttributeType_t attributeType,
+                                                    int64_t elementCount,
+                                                    void* arrayOfElements)
+{
+    return miopen::try_(
+        [&] {
+            miopen::backend_api::BackendDescriptor* theDescriptor =
+                static_cast<miopen::backend_api::BackendDescriptor*>(descriptor);
+
+            miopenStatus_t status = theDescriptor->SetAttribute(
```

Comment:
I'm not sure if every method on the internal class should return `miopenStatus_t`.